Define methods for big(T)

giordano opened the pull request:
This PR defines methods for `big(T)`, which return an appropriate type to represent a value of type `T` as an arbitrary precision number.  I believe this is useful *per se* and also to define a single simple method to promote a complex number to arbitrary precision, as suggested [here](https://github.com/JuliaLang/julia/issues/21204#issuecomment-289910964).  If this PR will be merged, I'll follow up with a patch implementing `big{T<:Real}(z::Complex{T})`.

giordano committed:
Define methods for big(T) and simplify big(::Complex)

StefanKarpinski commented:
@JeffBezanson has sometimes had objections to functions that work in both the type and value domain, but this one seems harmless enough to me – hard to imagine what the ambiguity would be.

StefanKarpinski commented:
Note that this needs to wait until after 0.6 is branched to be merged.

giordano commented:
Rebased on master, used new `where` syntax, added the method
```julia
big(z::Complex{T}) where {T<:Real} = Complex{big(T)}(z)
```
that replaces the two methods
```julia
big(z::Complex{<:AbstractFloat}) = Complex{BigFloat}(z)
big(z::Complex{<:Integer}) = Complex{BigInt}(z)
```
as suggested by @ararslan at https://github.com/JuliaLang/julia/issues/21204#issuecomment-289910964.

ararslan commented on base/complex.jl:
```diff
@@ -880,8 +880,8 @@ end
 
-big(z::Complex{<:AbstractFloat}) = Complex{BigFloat}(z)
-big(z::Complex{<:Integer}) = Complex{BigInt}(z)
+big(::Type{Complex{T}}) where {T} = Complex{big(T)}
```
It might make sense to impose `T<:Real` here.

giordano replied:
Done, thanks.

ararslan commented on base/number.jl:
```diff
@@ -211,3 +211,22 @@ julia> factorial(big(21))
+"""
+    big(T::Type)
+
+Returns an appropriate type to represent a value of type `T` as an arbitrary precision number.
```
I'd word this like
> Compute the type that represents the numeric type `T` with arbitrary precision.

giordano replied:
Reworded, thanks again for the suggestion!

giordano committed:
[skip ci] More clear docstring for big(T)

ararslan reviewed (approved):
Nice work!
